dev-docs/RFCs/v7.1/shader-module-injection-rfc.md

```diff
@@ -0,0 +1,71 @@
+# RFC: Shader Module Code Injection
+
+* **Author**: Tarek Sherif
+* **Date**: May, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC specifies a mechanism to allow shader modules to inject custom code into the source they are augmenting based on a defined set of directives.
+
+
+## Background
+
+Shader modules currently only add variables and define functions in shaders. Without the ability to inject code in the `main` function, shaders that use them are forced modify their code to use the functionality that a shader module provides.
+
+There is a separate code-injection mechanism in `BaseModel` that allows for code injection by the application. It implements most of the functionality required for module code injection, but limits injections to variable declaration segments, or the beginning or end of the `main` function. This leads to a relatively brittle system, where the application must write arbitrary GLSL code against shader code that might change at some point, and that it cannot affect any intermediate calculations.
```

Comment:
- The shader injection is in shadertools `assembleShaders` function. 
- `BaseModel` should just  be forwarding arguments to `assembleShaders` as a convenience.

Separate this paragraph into two, one that talks about `BaseModel`, and one that talks about limitations in the current shadertools injection system.

I think the current injection system also supports arbitrary string matching. Still brittle, but a bit more flexible than stated here.